tin/apps/courses/models.py

```diff
@@ -11,17 +13,31 @@ def filter_visible(self, user):
         if user.is_superuser:
             return self.all()
         else:
-            return self.filter(Q(teacher=user) | Q(students=user)).distinct()
+            return self.filter(
+                Q(teacher=user)
+                | (Q(students=user) & (Q(archived=False) | Q(permission__contains="r")))
+            ).distinct()
 
     def filter_editable(self, user):
         if user.is_superuser:
             return self.all()
         else:
             return self.filter(teacher=user).distinct()
 
+    def filter_permission(self, user, perm: Literal["-", "r", "w", "rw"]):
+        if user.is_superuser:
+            return self.all()
+        else:
+            return self.filter(Q(archived=False) | Q(permission__icontains=perm) | Q(teacher=user))
+
 
 class Course(models.Model):
     SORT_BY = (("due_date", "Due Date"), ("name", "Name"))
+    PERMISSIONS = (
+        ("rw", "view and submit assignments"),
+        ("r", "view assignments"),
+        ("-", "see nothing"),
+    )
```

Comment:
Normally, for choice tuples, the Tin codebase tries to keep the database value a single character. Plus, I don't like the `-`...

Maybe `S` for submittable, `V` for viewable, and `N` for nothing?

This is a topic of convention and I'm not entirely sure what the right way is, but since these aren't shown to the user and are (hopefully) replaced with non-icontains queries above, I think single-character database values would be better. Thoughts?

Reply:
I made the options `-`, `r`, and `w`. I don't like `S`, `V` and `N` because it's harder to tell what they mean.